### ISSUE TITLE
Query functions returning ranges?

### ISSUE THREAD

**Issue by mreineck** [healpy/healpy#641]
Unless I have missed something, there doesn't seem to be a variant of the pixel querying routines in `healpy` which return pixel ranges instead of individual pixel indices. At high resolutions this becomes quite a waste of resources, because it increases the memory consumption (as well as the run time) from `O(sqrt(Npix))` to `O(Npix)`, where `Npix` is the number of pixels inside the queried shape.

Since Healpix C++ already supports queries that return ranges, I don't think this would be hard to add to `healpy`; but I'm not sure what the user interface should look like (new function, new option to the existing routines, etc.).

**Comment by zonca** [comment]
Thanks, Can you please provide an example of this in C++?

**Comment by mreineck** [comment]
You are already using one :-)
https://github.com/healpy/healpy/blob/f6c408c418b1e4b56393ad77c5bc32014421dcf0/healpy/src/_query_disc.pyx#L63

One way to represent ranges in Python is to return arrays with an additional dimension of length 2 (for start and end of the range). My gut feeling is that this should be the last dimension, but that's just a personal opinion.

**Comment by zonca** [comment]
so you would want to avoid doing `pixset_to_array` and instead return the ranges directly?

**Comment by mreineck** [comment]
Yes, at least as an option. For queries that return a large number of pixels, the `pixset_to_array` part dominates the execution time and also allocates huge arrays, which would not be necessary when working on ranges only.

**Comment by zonca** [comment]
ok, so it should return a `pixset` numpy 2D array with shape `(num_ranges, 2)` where first column is start of the range, second column is the end of the range. It would be great if someone else wants to implement this.

**Comment by hivon** [comment]
This is mostly relevant for RING indexing applications (because the query_disc operates in the RING scheme).
In the NESTED case, going through the full list of pixels will remain necessary, even if only a rangeset is expected.

**Comment by mreineck** [comment]
Actually Healpix C++ has a recursive query_disc algorithm for the NESTED scheme which directly works on ranges too. It basically starts with the 12 base pixels and recursively subdivides those where it cannot determine that they are fully inside or outside the disk.
